core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala

```diff
@@ -207,18 +166,14 @@ private[spark] object HadoopFSUtils extends Logging {
     // Note that statuses only include FileStatus for the files and dirs directly under path,
     // and does not include anything else recursively.
     val statuses: Array[FileStatus] = try {
-      fs match {
-        // DistributedFileSystem overrides listLocatedStatus to make 1 single call to namenode
-        // to retrieve the file status with the file block location. The reason to still fallback
-        // to listStatus is because the default implementation would potentially throw a
-        // FileNotFoundException which is better handled by doing the lookups manually below.
-        case (_: DistributedFileSystem | _: ViewFileSystem) if !ignoreLocality =>
-          val remoteIter = fs.listLocatedStatus(path)
-          new Iterator[LocatedFileStatus]() {
-            def next(): LocatedFileStatus = remoteIter.next
-            def hasNext(): Boolean = remoteIter.hasNext
-          }.toArray
-        case _ => fs.listStatus(path)
+      if (ignoreLocality) {
+        fs.listStatus(path)
+      } else {
+        val remoteIter = fs.listLocatedStatus(path)
```

Comment:
yeah a FS can choose not to implement it (although all the main ones override this). If not implemented it will fall back to the default impl in `FileSystem`, which basically calls `listStatus` and then `getFileBlockLocations` on each `FileStatus` received. The behavior is very similar to what this class is doing later on.

Reply:
HDFS and S3A both do this; ABFS merits minor optimisation too. Because they return a remote iterator they can do paged fetch of data
* HDFS/webHDFS: paged download for better scalability
* S3A (3.3.1+): async prefetch of next page of data

ABFS should copy the S3A approach; it's listing API is paged too. 

Better to rely on the FS to do the work *but make clear you expect the maintainers to do so*

Reply:
Thanks @steveloughran , yes I also think it's better to rely on the FileSystem-specific `listLocatedStatus` impl rather than having the logic here. However, the change seems to break a few assumptions in the test cases so I'll isolate it into another PR.